File: ode-solvers/poisson_2D.f90
!=================================================================
! POISSON EQUATION SOLVER ALGORITHM
!=================================================================
! Author: Daniel Noal Pineda
! Email : noaldaniel41@gmail.com
! Date  : 2025
! Repository: https://github.com/tuusuario/tu-repo
!=================================================================
! OBJECTIVES: Solve the 2D Poisson equation using Jacobi, 
!             Gauss-Seidel, and Successive Over-Relaxation (SOR) methods
!
! NOTES: The method used is controlled by the variable icontrol
!        For correct operation, must be implemented within a CONTAINS 
!        block (does not work properly at the end of the file)
!
! INPUTS:
!         ·icontrol     : method selector (1=Jacobi, 2=Gauss-Seidel, 3=SOR)
!         ·T_cinicials  : initial temperature matrix
!         ·p            : source term matrix
!         ·Lx           : domain length in x-direction
!         ·Ly           : domain length in y-direction
!         ·h            : grid spacing
!         ·punt_i       : i-index for convergence monitoring point
!         ·punt_j       : j-index for convergence monitoring point
!         ·omega        : relaxation parameter (for SOR method)
!         ·unit_file    : file unit for writing convergence data
!
! OUTPUTS:
!         ·T_final      : final temperature matrix (solution)
!         ·conv_p       : convergence vector at the monitoring point
!=================================================================

SUBROUTINE SOLVER_POISSON_2D(icontrol,T_cinicials,T_final,p,Lx,Ly,h,punt_i,punt_j,conv_p,omega,unit_file)

    IMPLICIT NONE

    ! Inputs
    integer, intent(in) :: icontrol, punt_i, punt_j, unit_file
    double precision, intent(in) :: T_cinicials(:,:), p(:,:)
    double precision, intent(in) :: Lx, Ly, h, omega

    ! Outputs
    double precision, intent(out) :: T_final(:,:), conv_p(:)

    ! Internal Variables
    logical :: convergencia
    integer :: i, j, Nx, Ny, iteracio, max_iter
    double precision :: Error, precisio
    double precision, allocatable :: T_k(:,:), T_k_plus_1(:,:) 

    ! Convergence parameters
    precisio = 1.0d-4
    max_iter = 100000
    
    ! Matrix dimensions
    Nx = size(T_cinicials, 1)
    Ny = size(T_cinicials, 2)

    ! Memory allocation
    allocate(T_k(Nx,Ny), T_k_plus_1(Nx,Ny))
    
    ! Initialization
    convergencia = .false.
    iteracio = 0
    
    ! Copy initial conditions
    T_k = T_cinicials
    T_k_plus_1 = T_k

    ! Iterate until convergence
    do while (.not. convergencia .and. iteracio < max_iter)
        
        iteracio = iteracio + 1

        !-------------------------------
        ! JACOBI METHOD (ICONTROL = 1)
        !-------------------------------

        if (icontrol == 1) then
            do i = 2, Nx-1
                do j = 2, Ny-1

                    ! Jacobi method algorithm
                    T_k_plus_1(i,j) = (T_k(i+1,j) + T_k(i-1,j) + &
                                        T_k(i,j+1) + T_k(i,j-1) + &
                                        h**2 * p(i,j)) / 4.0d0

                enddo
            enddo

        !-------------------------------------
        ! GAUSS-SEIDEL METHOD (ICONTROL = 2)
        !-------------------------------------

        elseif (icontrol == 2) then
            do i = 2, Nx-1
                do j = 2, Ny-1

                    ! Gauss-Seidel algorithm
                    T_k_plus_1(i,j) = (T_k_plus_1(i-1,j) + T_k(i+1,j) + &
                                      T_k_plus_1(i,j-1) + T_k(i,j+1) + &
                                      h**2 * p(i,j)) / 4.0d0

                enddo
            enddo

        !---------------------------------------
        ! SOR METHOD (ICONTROL = 3)
        !---------------------------------------

        elseif (icontrol == 3) then
            do i = 2, Nx-1
                do j = 2, Ny-1

                    ! Successive Over-Relaxation (SOR) algorithm
                    T_k_plus_1(i,j) = T_k(i,j) + omega * (T_k_plus_1(i-1,j) + T_k(i+1,j) + &
                                      T_k_plus_1(i,j-1) + T_k(i,j+1) - 4*T_k(i,j) + &
                                      h**2 * p(i,j)) / 4.0d0

                enddo 
            enddo
        endif

        !------------------------------------------------
        ! CONVERGENCE - ERROR CALCULATION - NEXT ITERATION
        !------------------------------------------------
        
        ! Study convergence at a point and save data to file
        write(unit_file, '(I20, F20.6)') iteracio, T_k_plus_1(punt_i,punt_j)

        ! Save to convergence vector
        conv_p(iteracio) = T_k_plus_1(punt_i,punt_j)

        ! Calculate the error
        Error = 0.0d0

        do i = 2, Nx-1
            do j = 2, Ny-1
                Error = max(Error, abs(T_k_plus_1(i,j) - T_k(i,j)))
            enddo
        enddo

        ! Check convergence
        if (Error < precisio) then
            convergencia = .TRUE.
        endif

        ! Prepare for the next iteration
        T_k = T_k_plus_1
    
    enddo

    ! Save the final result
    T_final = T_k_plus_1
    
    deallocate(T_k, T_k_plus_1)

    RETURN

END SUBROUTINE SOLVER_POISSON_2D

File: Python/ode-solvers/poisson_2D.py
#=================================================================
# POISSON EQUATION SOLVER ALGORITHM
#=================================================================
# Author: Daniel Noal Pineda
# Email : noaldaniel41@gmail.com
# Date  : 2026
# Repository: https://github.com/Dnoal17/numerical-methods.git
#=================================================================
# OBJECTIVES: Solve the Heat Equation (2D Poisson equation) using Jacobi, 
#             Gauss-Seidel, and Successive Over-Relaxation (SOR) methods
#
# NOTES: The method used is controlled by the variable icontrol
# 
# INPUTS:
#         ·icontrol     : method selector (1=Jacobi, 2=Gauss-Seidel, 3=SOR)
#         ·t_cinitials  : initial temperature matrix
#         ·p            : source term matrix
#         ·h            : grid spacing
#         ·i_point      : i-index for convergence monitoring point
#         ·j_point      : j-index for convergence monitoring point
#         ·omega        : relaxation parameter (for SOR method)
#         ·unit_file    : file unit for writing convergence data
#
# OUTPUTS:
#         ·T_final      : final temperature matrix (solution)
#         ·conv_p       : convergence vector at the monitoring point
#=================================================================

import numpy as np

def solver_poisson_2d(icontrol, t_cinitials, p, h, i_point, j_point, omega):

    # Convergence parameters
    precision = 1.0e-4
    max_iter = 100000

    # Matrix dimensions
    Nx, Ny = t_cinitials.shape

    # Initialize convergence vector
    conv_p = np.zeros(max_iter)

    # Initialization
    convergence = False
    iteration = 0

    # Copy initial conditions
    T_k = t_cinitials.copy()
    T_k_plus_1 = T_k.copy()

    # Iterate until convergence
    while not convergence and iteration < max_iter:

        iteration += 1

        #-------------------------------
        # JACOBI METHOD (ICONTROL = 1)
        #-------------------------------

        if icontrol == 1:

            # Jacobi method algorithm
            T_k_plus_1[1:-1, 1:-1] = (T_k[2:, 1:-1] + T_k[:-2, 1:-1] + T_k[1:-1, 2:] +
                T_k[1:-1, :-2] +  h**2 * p[1:-1, 1:-1]) / 4.0

        #-------------------------------------
        # GAUSS-SEIDEL METHOD (ICONTROL = 2)
        #-------------------------------------

        elif icontrol == 2:

            for i in range(1, Nx - 1):
                for j in range(1, Ny - 1):

                    # Gauss-Seidel algorithm
                    T_k_plus_1[i, j] = (T_k_plus_1[i - 1, j] + T_k[i + 1, j] +
                        T_k_plus_1[i, j - 1] + T_k[i, j + 1] + h**2 * p[i, j]) / 4.0

        #---------------------------------------
        # SOR METHOD (ICONTROL = 3)
        #---------------------------------------

        elif icontrol == 3:

            for i in range(1, Nx - 1):
                for j in range(1, Ny - 1):

                    # Successive Over-Relaxation (SOR) algorithm
                    T_k_plus_1[i, j] = T_k[i, j] + omega * (T_k_plus_1[i - 1, j] +
                        T_k[i + 1, j] + T_k_plus_1[i, j - 1] +T_k[i, j + 1] - 4.0 * T_k[i, j] +
                        h**2 * p[i, j]) / 4.0
        
        # Error if icontrol != 1,2,3
        else:

            print(f"ERROR: Icontrol = {icontrol}. Its value must be 1, 2, 3")
            return None

        #------------------------------------------------
        # CONVERGENCE - ERROR CALCULATION - NEXT ITERATION
        #------------------------------------------------

        # Save to convergence vector
        conv_p[iteration - 1] = T_k_plus_1[i_point, j_point]

        # Calculate the error
        error = np.max(np.abs(T_k_plus_1[1:-1, 1:-1] - T_k[1:-1, 1:-1]))

        # Check convergence
        if error < precision:
            convergence = True

        # Prepare the next iteration
        T_k = T_k_plus_1.copy()

    # Save the final result
    T_final = T_k_plus_1.copy()

    return T_final, conv_p[:iteration]